### PR TITLE
Adding comm_info_request and comm_info_reply messages

### PR DESCRIPTION
**Pull request by SylvainCorlay**
Adding a `comm_list_[request/reply]` shell message. This requires https://github.com/jupyter/jupyter_client/pull/34.

### CODE REVIEW

File: ipykernel/kernelbase.py [hunk 1]
```diff
@@ -470,6 +470,19 @@ def kernel_info_request(self, stream, ident, parent):
                                 self.kernel_info, parent, ident)
         self.log.debug("%s", msg)
 
+    def comm_list_request(self, stream, ident, parent):
+        content = parent['content']
+
+        # Should this be moved to ipkernel?
+        if hasattr(self, 'comm_manager'):
+            comm_list = self.comm_manager.comms.keys()
```
Comment (minrk) [inline ipykernel/kernelbase.py]: This will presumably need to be cast to a list.

Comment (SylvainCorlay) [inline ipykernel/kernelbase.py]: that's right. Python 3 sorry.

**Comment by SylvainCorlay** [comment]
Done

File: ipykernel/kernelbase.py [hunk 2]
```diff
@@ -470,6 +470,19 @@ def kernel_info_request(self, stream, ident, parent):
                                 self.kernel_info, parent, ident)
         self.log.debug("%s", msg)
 
+    def comm_list_request(self, stream, ident, parent):
+        content = parent['content']
+
+        # Should this be moved to ipkernel?
```
Comment (jasongrout) [inline ipykernel/kernelbase.py]: ipykernel

Comment (SylvainCorlay) [inline ipykernel/kernelbase.py]: I think that it is @minrk call.

Comment (minrk) [inline ipykernel/kernelbase.py]: I think this is a fine place for it.

**Comment by SylvainCorlay** [comment]
Rebased to use the new schema for the `comm_info_reply`. 

```yml
Message type: ``comm_info_request``::

    content = {
    }

Message type: ``comm_info_reply``::

    content = {
        # A dictionary of the comms, indexed by uuids.
        'comms': {
            comm_id: {
                'target_name': str,
            },
        },
    }
```

**Comment by SylvainCorlay** [comment]
I added the ability to filter by target name as discussed earlier.

**Comment by minrk** [comment]
thanks, @SylvainCorlay!

**Comment by SylvainCorlay** [comment]
I hesitated to give a *list* of target names rather than a single target name for the filter, but it would have been the only message having a list of target names, so it is probably more consistent with the rest of the messages to only allow one.

**Comment by minrk** [comment]
@SylvainCorlay good choice. I don't imagine the use case for people wanting more than one but less than all is worth having the extra logic.

**Comment by SylvainCorlay** [comment]
How can we move forward with this? Should there be a release of jupyter-client?

File: ipykernel/tests/test_message_spec.py [hunk 3]
```diff
@@ -419,6 +423,14 @@ def test_kernel_info_request():
     validate_message(reply, 'kernel_info_reply', msg_id)
 
 
+def test_comm_info_request():
+    flush_channels()
+
+    msg_id = KC.comm_info()
```
Comment (minrk) [inline ipykernel/tests/test_message_spec.py]: Can you add a `if not hasattr(KC, 'comm_info'): raise SkipTest()`? Then this test will run while comm_info is defined, but skip while jupyter_client is < 4.1.

Comment (SylvainCorlay) [inline ipykernel/tests/test_message_spec.py]: Done!

**Comment by SylvainCorlay** [comment]
(Done)

**Comment by minrk** [comment]
@SylvainCorlay thanks! Now I think we can start getting this tested for those living on master.

**Pull request merged**